Support for more services -- SageMaker Notebook and Gateway Endpoints (S3/Dynamo)

elementalvoid opened the pull request:
Fixes #49.

During a recent project my team @Ibotta found a couple of things that we wanted to adjust for our use case.

First, I found that the module doesn't support SageMaker Notebook because this service uses a "non-standard" fully qualified service name. It also doesn't seem to have a common name; it can only be referenced by the full name (ex: `aws.sagemaker.us-east-1.notebook`).

Second, I wanted to be able to use the module for both `Interface` and `Gateway` type Endpoints.

Lastly, I wanted to use for_each on the endpoints to avoid list sorting issues as we add/remove endpoints.

elementalvoid commented:
I had performed some manual checks against the different create/no-create options and full/empty lists during development. I noticed the automated tests aren't part of any CI so I'm running those right now.

elementalvoid commented:
Hoookay. Running those tests picked up a couple of bugs. I'm fixing and will update soon.

lorengordon commented:
Yay tests! If you have `go` installed, you should be able to run `make test` and it will loop through all the configs in `tests/`, applying and destroying. I'll be doing that to confirm things. Or if you don't have go you can init/apply/destroy manually in each directory for sure.

elementalvoid commented:
Yep! I've used terratest in other projects so I'm set there.

I did run down a rabbit hole for a bit because what _had worked_ manually all of a sudden didn't work via terratest! Turns out I was using a new TF version in my work project but my "default" version from `asdf` was older and had trouble mapping the `var.vpc_endpoint_services` list as an actual list. It considered it a `dynamic` type and that caused the `for_each = toset(..)` to fail because it wasn't properly mapping the variable's type. Sigh.

So, it fails with `0.12.8`. It passes with `0.12.24`. I yet attempted to bisect where the fail/pass changes.

What's your thought on requiring a greater minimum 0.12 patch version? Something like `>= 0.12.14` (once I find the breaking point).

elementalvoid commented:
I'm also now adding tests for the new `Gateway` Endpoint and the fully qualified name support. And, total honesty: I relied too heavily on `terraform plan` but I discovered via these tests that `Gateway` Endpoints are not allowed to specify Subnets. I'll have to rework that a bit.

No 🥇 for me. I know better than to trust terraform plans : )

elementalvoid commented:
I narrowed down the version problem for the `for_each` problem.

Terraform `0.12.8` produces:
```
go test -run TestModule/no_endpoints
# ... snip ...
TestModule/no_endpoints 2020-04-02T19:49:02-06:00 retry.go:72: terraform [apply -input=false -auto-approve -lock=false -no-color]
TestModule/no_endpoints 2020-04-02T19:49:02-06:00 command.go:87: Running command terraform with args [apply -input=false -auto-approve -lock=false -no-color]
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158: module.config_endpoint.data.aws_region.selected: Refreshing state...
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158: Error: Invalid for_each set argument
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:   on ../../main.tf line 10, in data "aws_vpc_endpoint_service" "this":
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:   10:   for_each = toset(var.create_vpc_endpoints ? var.vpc_endpoint_services : [])
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158: The given "for_each" argument value is unsuitable: "for_each" supports maps
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158: and sets of strings, but you have provided a set containing type dynamic.
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 command.go:158:
TestModule/no_endpoints 2020-04-02T19:49:08-06:00 retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
```

But with a single patch version update to Terraform `0.12.9` passes:
```
go test -run TestModule/no_endpoints
# ... snip ...
PASS
ok      terraform-aws-tardigrade-vpc-endpoints/tests    37.521s
```

So we'd need to bump `versions.tf` to:
```hcl
terraform {
  required_version = ">= 0.12.9"
}
```

lorengordon commented:
No objections to bumping the terraform version requirement. Please just add a short comment describing the problem with the earlier version...

elementalvoid commented on README.md:
```diff
@@ -2,6 +2,20 @@
+version `0.12.9` -- see the [Changelog here](https://github.com/hashicorp/terraform/blob/v0.12/CHANGELOG.md#0129-september-17-2019).
+The original [bug was reported as issue #22281](https://github.com/hashicorp/terraform/issues/22281).
+
+## Updating documentation
```
@lorengordon -- Let me know if this shouldn't be included here. I fumbled with Travis for a bit because I wasn't used to how your team's projects work.

lorengordon replied:
Sorry about that, I would have fixed the docs for you. Timing-wise, you got side-swiped by yesterday's terraform-docs update to 0.9.0 where they added the Requirements section. So then the presence of the diff caused the CI failure.

Things get a little touchy to fix yourself, because it's dependent on your local system. Though you clearly figured it out. Either of these should also have worked:

```
## This will install terraform-docs to your local system, may not
## be desirable for you. And may not work if the make target does
## not account for your OS
make docs/generate

## This will run terraform-docs in the docker container, which of
## course requires that you have docker...
make init  # initialize one time...
make docker/run target=docs/generate
```

lorengordon replied:
And our Makefiles utilize features of `make` that are only present in fairly recent versions, and not all OS platforms have packages for the latest versions. 🤕 Good times.

elementalvoid replied:
'tis all good. I was tempted to update the Dockerfile to the latest base image -- but not being familiar with the status of the `tardigrade-ci` project I chose not to.

I'll update the readme to use the make targets!

elementalvoid commented:
Thanks for bearing with me while I fixed up tests and docs for the CI checks.

I've added tests for both Gateway Endpoints and fully qualified names (SageMaker Notebook). I also added a test for per-Endpoint SecurityGroups.

All tests are passing in my AWS Sandbox account.

lorengordon commented:
@elementalvoid Thanks so much for the contribution! This is released in [`3.0.0`](http://github.com/plus3it/terraform-aws-tardigrade-vpc-endpoints/pull/52). Tag will show up very shortly...
